lib/drivers/ism43340/ism43340.cpp

```diff
@@ -0,0 +1,270 @@
+#include "Assert.hpp"
+#include "Logger.hpp"
+
+#include <memory>
+
+#define CMD_SET_HUMAN_READABLE "$$$"
+#define CMD_SET_MACHINE_READABLE "---"
+#define CMD_RESET_SETTINGS "Z0"
+
+#define CMD_JOIN_NETWORK "C0"
+#define CMD_SET_SSID "C1="
+#define CMD_SET_PASSWORD "C2="
+#define CMD_SET_SECURITY "C3="
+#define CMD_SET_DHCP "C4="
+#define CMD_GET_CONNECTION_INFO "C?"
+
+#define CMD_SET_TRANSPORT_PROTOCOL "P1="
+#define CMD_SET_HOST_IP "P3="
+#define CMD_SET_PORT "P4="
+#define CMD_START_CLIENT "P6="
+
+#define CMD_SEND_DATA "S3="
+
+Ism43340::Ism43340(SpiPtrT sharedSPI, PinName nCs, PinName intPin,
+                   PinName _nReset)
+    : CommLink(sharedSPI, nCs, intPin), nReset(_nReset) {
+    reset();
+}
+
+inline void Ism43340::formatCommand(std::string* command) {
+    if (command->size() % 2 == 0) {
+        command->append("\r\n");
+    } else {
+        command->append("\r");
+    }
+}
+
+int Ism43340::writetospi(std::string* command) {
+    while (dataready.read() != 1) {
+    }
+
+    uint8_t lastc = 0;
+    radiospi.chipselect();
+    for (int i = 0; i < command->size(); i++) {
+        uint8_t c = command[i];
+        if (lastc != 0) {
```

Comment:
Is there ever a chance you can have '\0' in the data being sent? This would fail in that case

Reply:
You can change the lastc to be signed and slightly larger, then make the default -1

Reply:
Yeah null bytes can def be a thing. We will need hard code max buffer length, and perhaps encode total length in the packet header and stop short based on the transmission length. Although, I don't anticipate the forward packets changing length, on the return when we wanna send telem